build-support/fetchgit: fails with repository containing submodules when enabling `leaveGitDir` and `fetchSubmodules`

### Describe the bug

`fetchgit` fails when using `leaveGitDir = true` and `fetchSubmodules = true` with repositories that contain submodules.

A checked-out repository that contains modules will have `.git` *files* pointing to `.git/modules/<submodule_path>`. git uses `.git` files to "link" submodules gitdirs to the root `.git/modules/<submodule>` path. I'm not sure if this is new git behavior and thus it broke `nix-prefetch-git`. See below in the reproduction section.

`nix-prefetch-git:clone_user_rev()` ([here](https://github.com/NixOS/nixpkgs/blob/67b08d46a1cf2fe7caa4439226f96a0434855b1b/pkgs/build-support/fetchgit/nix-prefetch-git#L327-L329)) tries to list `.git` directories and get their parent paths with `readlink(1)`, but finds `.git` files and `readlink` fails. [Because it's invoked in a command substitution as an argument for another command](https://www.shellcheck.net/wiki/SC2312), it fails silently and won't return an error to the parent shell (and won't trigger `errexit`), so in these cases `make_deterministic_repo ""` is called and fails with "not a git repository" errors.

A possible fix is to use `find "$dir" -type d -name ".git"` to filter directories. I believe this option/value is portable. This will only give `readlink` directory paths and avoid the very subtly error.

### Steps To Reproduce

Checked-out repository with submodule:

```console
> cd "$(mktemp -d)"
> git init
Initialized empty Git repository in /tmp/tmp.rQmCGRJ5ey/.git/
> git submodule add --depth 1 https://github.com/NixOS/rfcs
> git submodule status
e364d5f0cddc03d164f0d7dfc8d6c4f1616364b7 rfcs (heads/master)
> stat rfcs/.git
  File: rfcs/.git
  Size: 29        	Blocks: 8          IO Block: 4096   regular file
  ...
> cat rfcs/.git
gitdir: ../.git/modules/rfcs
```

Emulating the `leaveDotGit = true` branch:

```console
> find "$PWD" -name .git | while read -r gitdir; do echo "make_deterministic_repo '$(readlink -f "$gitdir/..")'"; done
make_deterministic_repo ''
make_deterministic_repo '/tmp/tmp.rQmCGRJ5ey'

> find "$PWD" -name .git
/tmp/tmp.rQmCGRJ5ey/rfcs/.git
/tmp/tmp.rQmCGRJ5ey/.git

> echo "$(readlink -f "/tmp/tmp.rQmCGRJ5ey/rfcs/.git/..")" | xxd
00000000: 0a                                       .
          ^ `echo` receives an empty string and just prints a newline

> readlink -f "/tmp/tmp.rQmCGRJ5ey/rfcs/.git/.."; echo $?
1
```

### Expected behavior

Expected `fetchgit` to correctly fetch a repository with submodules and leave `.git` directories intact when `leaveDotGit` is true.

### Notify maintainers

I'm not sure who are `fetchgit` owners. I'm tagging the last 3 who touched `nix-prefetch-git`: @marsam @domenkozar @fetsorn

### Metadata

```console
> nix shell nixpkgs#nix-info -c nix-info -m
 - system: `"x86_64-linux"`
 - host os: `Linux 6.3.8-100.fc37.x86_64, Fedora Linux, 37 (KDE Plasma), nobuild`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.15.1`
 - channels(root): `"nixpkgs"`
 - channels(k): `"nixpkgs"`
 - nixpkgs: `/home/k/.nix-defexpr/channels/nixpkgs`
```

--- a/pkgs/build-support/fetchgit/nix-prefetch-git
+++ b/pkgs/build-support/fetchgit/nix-prefetch-git
@@ -256,9 +256,15 @@ make_deterministic_repo(){
     cd "$repo"
     # Remove files that contain timestamps or otherwise have non-deterministic
     # properties.
-    rm -rf .git/logs/ .git/hooks/ .git/index .git/FETCH_HEAD .git/ORIG_HEAD \
-        .git/refs/remotes/origin/HEAD .git/config
-
+    if [ -f .git ]; then
+	local dotgit_content=$(<.git)
+	local dotgit_dir="${dotgit_content#gitdir: }"
+    else
+	local dotgit_dir=".git"
+    fi
+    pushd "$dotgit_dir"
+        rm -rf logs/ hooks/ index FETCH_HEAD ORIG_HEAD refs/remotes/origin/HEAD config
+    popd
     # Remove all remote branches.
     git branch -r | while read -r branch; do
         clean_git branch -rD "$branch"
@@ -276,7 +282,7 @@ make_deterministic_repo(){
     # Do a full repack. Must run single-threaded, or else we lose determinism.
     clean_git config pack.threads 1
     clean_git repack -A -d -f
-    rm -f .git/config
+    rm -f "$dotgit_dir/config"
 
     # Garbage collect unreferenced objects.
     # Note: --keep-largest-pack prevents non-deterministic ordering of packs
@@ -325,7 +331,7 @@ clone_user_rev() {
         find "$dir" -name .git -print0 | xargs -0 rm -rf
     else
         find "$dir" -name .git | while read -r gitdir; do
-            make_deterministic_repo "$(readlink -f "$gitdir/..")"
+            make_deterministic_repo "$(readlink -f "$(dirname "$gitdir")")"
         done
     fi
 }

